### PR TITLE
Building an Appointment Scheduling Telegram Bot with Python and Fauna

### PR DESCRIPTION
**Pull request by Chukslord1**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by lalith1403** [comment]
Do go through the checklist and tick whichever ones are applicable.

**Comment by ahmadmardeni1** [comment]
I will review this one.
cc @hectorkambow

**Comment by ahmadmardeni1** [comment]
Hello @Chukslord1  👋👀
At a quick glance, this seems to be an interesting article. Please start with the following:
1- Cut sentences down, there are a lot of unnecessarily long sentences. 
Long sentences will decrease the readability. [This tool](https://hemingwayapp.com/) may help you.
2- I found some grammar mistakes. Please use [this tool](https://app.grammarly.com/) and [this one](https://prowritingaid.com/).
3- You didn't use our preferred heading structure. [H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers). Edit all headers.
4- Link the Prerequisites to their official websites. Like: [Telegram](https://telegram.org/).
5- Put a dot at the end of every line. (I found some lines without dots).


Refer to the content published earlier to get a feel for the type of content and explanations we expect from the authors.
Please ping me when you are done so we can continue with the revision process.

**Comment by Chukslord1** [comment]
@ahmadmardeni1 I have made the changes required, thank you

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 1]
```diff
@@ -0,0 +1,514 @@
+Have you ever wanted to build a telegram bot that would allow you to schedule or plan your appointments? If your answer is yes, then this article is just what you need. You would also be utilizing a serverless database system called FaunaDB to build our system which will make your work easier. Now that you know what this article entails, let's get to the building part.
+
+Not so many people know about Fauna and how amazing it can be when managing databases. Therefore, I will be doing a little introduction to the Fauna serverless database system.
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:4]: Add Introduction header.
Delete the last line.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+### What is Fauna?
+
+Fauna is a client-side serverless document database that uses GraphQL and the Fauna Query Language (FQL) to support various data types and relational databases in a serverless API. You can learn more about Fauna in their official documentation [here](https://docs.fauna.com/fauna/current/).
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:8]: Delete "You can learn more about Fauna in their official documentation" and add that link to the first word "[Fauna](https://docs.fauna.com/fauna/current/)".

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+### Creating A Fauna Database
+
+To make use of Fauna, you have to signup on their website [here](https://dashboard.fauna.com/accounts/register). After signing up, you can create a new database by clicking on the `CREATE DATABASE` button on your dashboard. Provide a name for your database and save it.
+
+![database_dashboard](/engineering-education/building-an-appointment-scheduling-telegram-bot/database_dashboard.png)
+
+#### Creating the Fauna Collections
+
+For this article, you will need to create two collections in your database. A collection is similar to tables in your conventional database system. The first collection you need to create is the `Users` collection. This is where you will save all the user’s information. The second collection you need to create is the `Appointments` collection. This is where you will create all the user’s appointments.
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:17]: For this tutorial, you will need to create two collections in your database. A collection is similar to [tables](https://en.wikipedia.org/wiki/Table_(database)#:~:text=A%20table%20is%20a%20collection,table%20format%20within%20a%20database.&text=In%20relational%20databases%2C%20and%20flat,a%20row%20and%20column%20intersect.) in a [relational database system](https://en.wikipedia.org/wiki/Relational_database). The first collection you need to create is the `Users` collection. This is where you will save all the user’s information. The second collection you need to create is the `Appointments` collection. This is where you will create all the user’s appointments.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+
+To create a collection, click on `CREATE COLLECTION` then provide information for the required fields. You will be required to enter the name of the collection, the `History Days` and `TTL`. The `History Days` field refers to the number of days you want Fauna to retain a historical record of any data in that collection. The `TTL` refers to an expiry date for data in the collection. For example, if the `TTL` is 3, Fauna will automatically delete any data stored in the database three (3) days after the last date you modified it.
+
+![create_collection](/engineering-education/building-an-appointment-scheduling-telegram-bot/create_collection.png)
+
+After saving the collections you just created, you will notice there are no documents in your collections. A document is similar to rows of data in a table as in your conventional database system.
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:24]: After saving the collections you just created, you will notice there are no documents in your collections. A document is similar to [rows](https://en.wikipedia.org/wiki/Row_(database)#:~:text=In%20the%20context%20of%20a,data%20item%20in%20a%20table.&text=Each%20row%20in%20a%20table,table%20has%20the%20same%20structure.) of data in a table as in a relational database system.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+#### Creating a Fauna Index
+
+You will need to create a Fauna index that will allow you to scroll through the data created in your database. To do this, go to the `DB Overview` tab on the left side of your screen, then click on the `New Index` button.
+
+For this article, you will be creating three indexes; `users_index`, `appointment_index`, and `appointment_today_index`. The `users_index` index will enable you to scroll through data in the `Users` collection using the `id` field as a parameter to match. The `appointment_index` index will enable you to scroll through data in the `Appointments` collection using the `user_id` field as a parameter to match. The `appointment_today_index` index will allow you to scroll through data in the `Appointments` collection. It will use the `user_id` and `date_due` fields as parameters to match.
+
+![create_index](/engineering-education/building-an-appointment-scheduling-telegram-bot/create_index.png)
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:32]: You are saying "for this article" a lot. Please change it. For example:
We will create three indexes...
and why there is a semicolon after "three indexes"?

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+#### Prerequisites
+
+From this point onwards, to follow this step by step guide on building an appointment scheduling telegram bot with fauna, you need to have the following installed:
+
+
+* [Python 3.7 or >3.7](http://python.org/)
+* [Faunadb](https://pypi.org/project/faunadb/)
+* [telegram](https://pypi.org/project/telegram/)
+* [python_telegram_bot](https://pypi.org/project/python-telegram-bot/)
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:61]: Move this section to the beginning of the article (after the introduction part).
Put a dot at the end of every line.
Change the first line to:
To follow along with this tutorial, you need to have:

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+print(indexes)
+
+
+```
+If your result after running is similar to the image below, then you are good to go.
+
+![run_result](/engineering-education/building-an-appointment-scheduling-telegram-bot/run_result.png)
+
+# Creating the Telegram Bot
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:101]: change the heading structure to ###.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+
+from faunadb import query as q
+
+from faunadb.objects import Ref
+
+from faunadb.client import FaunaClient
+
+```
+
+We are importing `telegram`, telegram libraries, the python `datetime` module, and `faunadb` libraries, all of which we will be utilizing while building the bot.
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:147]: No need for this line. Delete it.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+
+Now you need to create a dispatcher and updater for your bot. Copy and paste the code below after your imports in your python file.
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:149]: Now let's create a dispatcher and an updater for your bot.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 10]
```diff
@@ -0,0 +1,514 @@
...
+
+updater = Updater(token=telegram_bot_token, use_context=True)
+
+dispatcher = updater.dispatcher
+
+```
+
+The dispatcher handles and processes the received message while the updater tracks, monitors, and reads messages sent to the bot and delivers them to the dispatcher.
+
+Now you need to write a script to start your bot. Copy and paste the code below in your python file.
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:165]: Now let's start our bot by:

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 11]
```diff
@@ -0,0 +1,514 @@
...
+
+![search_bot](/engineering-education/building-an-appointment-scheduling-telegram-bot/search_bot.png)
+
+Open your bot and start it by typing in the `/start` command. You will now receive a welcome message from the bot, as seen in the image below:
+
+![start_bot](/engineering-education/building-an-appointment-scheduling-telegram-bot/start_bot.png)
+
+#### Creating New Appointments
+
+To enable your telegram bot to create new appointments in the fauna database, copy and paste the code below in your python file.
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:247]: Now we will enable your telegram bot to create new appointments in the fauna database.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 12]
```diff
@@ -0,0 +1,514 @@
...
+
+    context.bot.send_message(chat_id=chat_id, text="Enter the appointment event you want to add along with its due in this format(mm/dd/yyyy) date separated by a comma 😁")
+
+```
+
+In the code above, you defined a method called `add_appointment` where you passed `update` and `context` as parameters. In the function, you saved the user’s message and `chat_id`. You then made use of the `chat_id` to query the FQL client using  the `users_index` index to check the user data that matches the `chat_id`.
+
+Next, you updated the `last_command` field in the user's data to save the `add_appointment` command the user sent. You then sent a message requesting the user to enter the name of the appointment and its due date separated by a comma.
+
+You have to create another method called `echo` that will collect the name and the due date for the new appointment then save it to the database. Copy and paste the code below after your `add_appointment` method in your python file.
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:268]: Now you have to create another method called `echo` that will collect the name and the due date for the new appointment then save it to the database.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 13]
```diff
@@ -0,0 +1,514 @@
...
+        context.bot.send_message(chat_id=chat_id, text="Successfully added appointment event 👍")
+
+```
+
+In the code above, you defined a method called `echo` where you passed `update` and `context` as parameters. In the function, you saved the user’s message and `chat_id`. You then made use of the `chat_id` to query the FQL client using the `users_index` index to check the user data that matches the `chat_id`.
+
+Next, you checked if the `last_command` in the data retrieved is an `add_appointment` command. If it is, you proceed to query the FQL client to create an appointment using the Fauna `create` method in the `Appointments` collection with the information provided in the user’s message.
+
+Finally, you need to create a command handler for the `add_appointment` command and a message handler for the `echo` method. The message handler filters the message and triggers the `echo` method when a message the user enters a message( a message is any text that does not start with `/`).
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:314]: Finally, you need to create a command handler for the `add_appointment` command and a message handler for the `echo` method. The message handler filters the message and triggers the `echo` method when the user enters a message (any text that does not start with `/`).

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+dispatcher.add_handler(MessageHandler(Filters.text, echo))
+
+```
+
+![add_appointment](/engineering-education/building-an-appointment-scheduling-telegram-bot/add_appointment.png)
+
+#### Listing Appointments
+
+To list all the appointments saved for a user, copy and paste the code below in your python file.
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:330]: The following part will provide the feature of listing all the appointments saved for a user.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 15]
```diff
@@ -0,0 +1,514 @@
...
+   if event_message == "":
+
+       event_message = "You dont hava any appointments saved, type /add_appointment to schedule one now 😇"
+
+   context.bot.send_message(chat_id=chat_id, text=event_message)
+
+```
+
+In the code above, you defined a method called `list_appointments` where you passed `update` and `context` as parameters. In the function, you saved the `chat_id` of the user and created an empty variable called `event_message`. You then made use of the `chat_id` to query the FQL client using the `appointments_index` index to check the user data that matches the `chat_id`.
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:365]: You added this part "In the code above, you defined a method called `list_appointments` where you passed `update` and `context` as parameters" after every part of code.
Please go through the entire article and delete it. (leave the first one)

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 16]
```diff
@@ -0,0 +1,514 @@
...
+       event_message = "You dont have any appointments saved, type /add_appointment to schedule one now 😇"
+
+   context.bot.send_message(chat_id=chat_id, text=event_message)
+
+
+```
+
+
+The `list_today_appointments` method is similar to the `list_appointments` method with one striking difference, the `appointment_today_index`. The `appointment_today_index` retrieves only data where the `date_due` field matches the current date.
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:412]: You duplicated "the `appointment_today_index`". Delete one.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 17]
```diff
@@ -0,0 +1,514 @@
...
+Copy and paste the code below to create a command handler for both methods above.
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:414]: Let's create a command handler for both methods.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 18]
```diff
@@ -0,0 +1,514 @@
...
+```
+
+![list_today_appointment](/engineering-education/building-an-appointment-scheduling-telegram-bot/list_today_appointment.png)
+
+#### Updating Appointments
+
+To enable your bot to update appointments in the database, copy and paste the code below in your python file.
+
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:429]: Now you will enable your bot to update appointments in the database.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 19]
```diff
@@ -0,0 +1,514 @@
...
+dispatcher.add_handler(MessageHandler(Filters.regex("/update_[0-9]*"), update_appointment))
+
+```
+
+![update_appointment](/engineering-education/building-an-appointment-scheduling-telegram-bot/update_appointment.png)
+
+#### Deleting Appointments
+
+To enable your bot to delete appointments in the database, copy and paste the code below in your python file.
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:477]: Same as the previous comments.

File: articles/building-an-appointment-scheduling-telegram-bot/index.md [hunk 20]
```diff
@@ -0,0 +1,514 @@
...
+```
+
+![delete_appointment](/engineering-education/building-an-appointment-scheduling-telegram-bot/delete_appointment.png)
+
+### Conclusion
+
+In this article, we built an appointment scheduling telegram bot with [Fauna's serverless database](https://fauna.com/). We saw how easy it is to integrate Fauna into a Python application and got the chance to explore some of its core features and functionalities.
+
+The source code of our bot is available on [Github](https://github.com/Chukslord1/FAUNA_APPOINTMENT_SCHEDULER_BOT). If you have any questions, don't hesitate to contact me on Twitter: [@](https://twitter.com/LordChuks3)[LordChuks3](https://twitter.com/LordChuks3).
+
```
Comment (ahmadmardeni1) [inline articles/building-an-appointment-scheduling-telegram-bot/index.md:513]: Delete the last part.

**Comment by ahmadmardeni1** [comment]
Good evening @Chukslord1👋
Still, there are a lot of unnecessarily long sentences. I have made some changes for your reference. Please go through the entire article and edit it. Also, there are some duplicated sentences. Please make sure to fix them.

**Comment by Chukslord1** [comment]
I have made the required changes, thanks for the review.

**Comment by Chukslord1** [comment]
@ahmadmardeni1 I really appreciate this review, feel free to let me know if there are other suggestions you have.

**Comment by ahmadmardeni1** [comment]
@Chukslord1 👋
Please go through the changes I made in f473f63 and 597882d and make sure to avoid those mistakes in future tutorials.
Keep up the wonderful work!✔️

**Comment by Chukslord1** [comment]
This fixes #1669

**Comment by hectorkambow** [comment]
[Building an Appointment Scheduling Telegram Bot with Python and Fauna #1828.pdf](https://github.com/section-engineering-education/engineering-education/files/6246375/Building.an.Appointment.Scheduling.Telegram.Bot.with.Python.and.Fauna.1828.pdf)


Upon running the article through our 3rd party plagiarism checker it raised a few red flags - & the % was higher than we typicaly like to accept. 

(see pdf)

**Comment by Chukslord1** [comment]
@hectorkambow I have successfully enhanced my article which contains less than 10%  plagiarism.

**Comment by ahmadmardeni1** [comment]
The tutorial should have everything that readers need. I can see there are a lot of links to other articles. Please fix it. @Chukslord1

**Comment by hectorkambow** [comment]
@Chukslord1  

I see his point - I believe he is thinking that leading out to another article may lead the reader to leaving the article - from that perspective it makes sense. 

I'd say as long as the _MAIN (topic) covered is covered completely within your article_ and we are only linking out for set up of basic boiler type code - or something basic like that. 

We would not want to link out to articles that cover the exact same topic we are writing about - because in that case - we wouldnt be writing/publishing the article. 

Hope that helps clear things up a bit - but great point Ahmad 🚀 

cc @ahmadmardeni1

**Comment by Chukslord1** [comment]
@ahmadmardeni1 Some instructions in my code are general to any article writing about the Fauna database, therefore, producing plagiarism issues.  I have relinked the instructions in a GitHub gist file instead of redirecting to other articles as instructed by @hectorkambow

**Comment by Chukslord1** [comment]
@hectorkambow the article is ready for the plagiarism check.

**Comment by hectorkambow** [comment]
plagiarism check done

**Pull request merged**